### PR TITLE
[java] Java 21 build support

### PR DESCRIPTION
**Pull request by Craigacp**
### Description
Bump spotless and the Gradle wrapper to 6.25.0 and 8.6 respectively to allow compiling ORT on Java 21. The build still targets Java 8.

I'm not sure if there will be CI changes necessary to use this PR, specifically for the Gradle version as I don't know if that is cached somewhere earlier in the CI build process.

The new Gradle version adds a warning that using `--source` and `--target` to select the Java language version is obsolete which is annoying, we can fix it if we decide to only allow building on newer versions of Java, while still supporting running on Java 8.

### Motivation and Context
Java 21 is the latest LTS release of Java and ORT should be able to build on it.

### CODE REVIEW

**Comment by Craigacp** [comment]
The spotless plugin now requires Java 11 to run, which is usually fine as that's the default JDK for the builds, however CodeQL is autodetecting that the project targets Java 8 and forcing the JDK to be Java 8 which breaks the code scanner as spotless can't run.

Looks like the fix involves an expansion of the CodeQL yaml - https://github.com/github/codeql/issues/13541#issuecomment-1607409543. Would it be ok to do that?

**Comment by snnn** [comment]
@Craigacp , there is a build failure in "CodeQL / Analyze (java)" build job. Would you mind taking a look?

**Comment by Craigacp** [comment]
> @Craigacp , there is a build failure in "CodeQL / Analyze (java)" build job. Would you mind taking a look?

Yeah, that's what my earlier comment is about (https://github.com/microsoft/onnxruntime/pull/19876#issuecomment-1994479938). The CodeQL autobuild system is autodetecting and then forcing Java 8 because we target that language level, so it's running the whole gradle process with Java 8, but the version of spotless that supports Java 21 requires at least Java 11 to run. The fix is to rewrite the CodeQL yaml file to split out the Java build separately from the rest and specify its environment, but I'm not that familiar with CodeQL configuration and so I'll likely need to run that job repeatedly to get the syntax right. As we run the CI builds with Java 11 by default it's not an issue for those builds, just the CodeQL pass.

There is a slightly bigger question which is due to that spotless change it'll be no longer possible to build ORT with Java 8 as the host JVM (though the generated jar file will still work fine on Java 8), and assuming we're ok with that it needs documenting.

**Comment by tianleiwu** [comment]
@justinchuby, could you help take a look how to make `CodeQL/Analyze (java)` target at Java 11?

**Comment by justinchuby** [comment]
> @justinchuby, could you help take a look how to make `CodeQL/Analyze (java)` target at Java 11?

I would love to take a look but I will need additional help as I don't have experience building java project. @Craigacp do have any clues about the build errors?

**Comment by justinchuby** [comment]
@Craigacp If I bump the current build tool chain to use java 21, will the main branch build?

**Comment by Craigacp** [comment]
Yes, it uses Java 11 for the builds anyway. It's just this one comes in a different path.

**Comment by justinchuby** [comment]
Testing with https://github.com/microsoft/onnxruntime/pull/19999

**Comment by Craigacp** [comment]
The macOS failure starts with an IOException but then gives an error about it not finding the new spotless. This runs fine on my M3 Mac using the gradle wrapper, so I'm not sure if it was a transient network issue which caused the IOException and then it complained, or if there's something different about the CI environment. Could someone restart that pipeline?

**Comment by tianleiwu** [comment]
MacOS pipeline failed:

[ 33%] Building CXX object CMakeFiles/onnxruntime_graph.dir/Users/runner/work/1/s/onnxruntime/core/graph/graph_proto_serializer.cc.o

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'onnxruntime'.
> Could not resolve all files for configuration ':classpath'.
   > Could not resolve com.diffplug.spotless:spotless-plugin-gradle:6.25.0.
     Required by:
         project : > com.diffplug.spotless:com.diffplug.spotless.gradle.plugin:6.25.0
      > No matching variant of com.diffplug.spotless:spotless-plugin-gradle:6.25.0 was found. The consumer was configured to find a library for use during runtime, compatible with Java 8, packaged as a jar, and its dependencies declared externally, as well as attribute 'org.gradle.plugin.api-version' with value '8.6' but:
          - Variant 'apiElements' capability com.diffplug.spotless:spotless-plugin-gradle:6.25.0 declares a library, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component for use during compile-time, compatible with Java 11 and the consumer needed a component for use during runtime, compatible with Java 8
              - Other compatible attribute:
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.6')
          - Variant 'javadocElements' capability com.diffplug.spotless:spotless-plugin-gradle:6.25.0 declares a component for use during runtime, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.6')
          - Variant 'runtimeElements' capability com.diffplug.spotless:spotless-plugin-gradle:6.25.0 declares a library for use during runtime, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component, compatible with Java 11 and the consumer needed a component, compatible with Java 8
              - Other compatible attribute:
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.6')
          - Variant 'sourcesElements' capability com.diffplug.spotless:spotless-plugin-gradle:6.25.0 declares a component for use during runtime, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.6')

**Comment by Craigacp** [comment]
Hmm. I'll try some more to replicate that locally. It's not failing on either of the macOS machines I have access to. I am having some difficulty compiling the C++ code but I think that's because I have XCode related issues on the newer machines.

**Comment by Craigacp** [comment]
Ahh. macOS is building it with Java 8:
```
Found Java: /Users/runner/hostedtoolcache/Java_Temurin-Hotspot_jdk/8.0.402-6/x64/Contents/Home/bin/java (found version "1.8.0.402")
```

**Comment by Craigacp** [comment]
Python looks like it uses `UsePythonVersion` from the Azure pipelines, so maybe we want to add the Java equivalent - https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/java-tool-installer-v0?view=azure-pipelines ? Do people at MS have opinions or I can hack on the pipeline some?

**Comment by edgchen1** [comment]
> Do people at MS have opinions or I can hack on the pipeline some?

sure, feel free

**Comment by tianleiwu** [comment]
Maybe insert a task for Java install here:
https://github.com/microsoft/onnxruntime/blob/ab2eaedfaa1e1abe5c128cfbccf8c6a45b75480e/tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml#L47

like
```
      - ${{ if eq(parameters['buildJava'], 'true') }}:
        - ${{ if eq(parameters['MacosArch'], 'x86_64') }}:
          - task: JavaToolInstaller@0
            inputs:
              versionSpec: "11"
              jdkArchitectureOption: "x64"
              jdkSourceOption: 'PreInstalled'
```

**Comment by Craigacp** [comment]
Ok, I added it. Not sure if it could be further minimized (maybe the two if statements on the architecture could be coalesced), but I think this should work.

File: tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml [hunk 1]
```diff
@@ -52,6 +52,19 @@ jobs:
     inputs:
       versionSpec: '18.x'
 
+  - ${{ if eq(parameters.MacosArch, 'x86_64') }}:
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml:55]: I believe the host architecture is that of macOS-latest, which is not arm64. I think the MacosArch parameter specifies the target architecture.

as this is specifying the Java on the host, could we always use x64?

Comment (Craigacp) [reply tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml:55]: Maybe, provided you didn't want to run the tests, as those will have an arm64 native binary in them and the x64 JVM won't load that. But I'm not clear on how the macOS arm64 support works in the CI pipelines.

Comment (edgchen1) [reply tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml:55]: Is running the tests possible, say for arm64 binaries on an x64 host?

FWIW, I think we only run tests when MacosArch is x86_64:
https://github.com/microsoft/onnxruntime/blob/b95fd4e644775a4343c13435bd729bd64f411752/tools/ci_build/github/azure-pipelines/templates/mac-cpu-packaging-steps.yml#L55-L58

AFAIK, none of the Mac CI agents we use now (all hosted agents) are arm64.

Comment (Craigacp) [reply tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml:55]: I don't think it's possible to run the arm64 tests on an x64 host without some kind of VM infrastructure. So I probably should tell it to load the x64 JVM everywhere. Presumably there's something in the CMake which forces the binaries to build as arm64 on x64, but nothing actually executes those binaries?

Comment (edgchen1) [reply tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml:55]: > Presumably there's something in the CMake which forces the binaries to build as arm64 on x64, but nothing actually executes those binaries?

Right, we can cross-compile. It looks like the build definition is skipping running of the tests for arm64.

The latest MacOS CI build succeeded. I think we can simplify this to just use the x64 JVM and it should be good.

Comment (Craigacp) [reply tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml]: Done.

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**